linker/shared/src/test/scala/org/scalajs/linker/testutils/IRAssertions.scala

```diff
@@ -0,0 +1,158 @@
+
+import scala.language.implicitConversions
+
+import scala.util.control.ControlThrowable
+
+import org.scalajs.ir.ClassKind
+import org.scalajs.ir.EntryPointsInfo
+import org.scalajs.ir.Definitions._
+import org.scalajs.ir.Traversers._
+import org.scalajs.ir.Trees._
+import org.scalajs.ir.Types._
+
+import org.scalajs.linker.standard._
+
+import org.junit.Assert._
+import org.scalajs.ir.Trees
+
+object IRAssertions {
+  implicit def classDefAssertions(classDef: ClassDef): ClassDefAssertions =
+    new ClassDefAssertions(classDef)
+
+  implicit def linkedClassAssertions(linkedClass: LinkedClass): LinkedClassAssertions =
+    new LinkedClassAssertions(linkedClass)
+
+  type Pat = PartialFunction[IRNode, Boolean]
+
+  private def patToTotal(pat: Pat): IRNode => Boolean =
+    node => pat.applyOrElse(node, (_: IRNode) => false)
+
+  abstract class AbstractIRNodeAssertions[T](selfNode: T) {
+    protected def newTraverser(f: IRNode => Boolean): TestTraverser[T]
+
+    def has(trgName: String)(pf: Pat): this.type = {
+      val tr = newTraverser(patToTotal(pf))
+      assertTrue(s"AST should have $trgName", tr.find(selfNode))
+      this
+    }
+
+    def hasNot(trgName: String)(pf: Pat): this.type = {
+      val tr = newTraverser(patToTotal(pf))
+      assertFalse(s"AST should not have $trgName", tr.find(selfNode))
+      this
+    }
+
+    def hasExactly(count: Int, trgName: String)(pf: Pat): this.type = {
+      var actualCount = 0
+      val tr = newTraverser(patToTotal(pf).andThen { matches =>
+        if (matches)
+          actualCount += 1
+        matches
+      })
+      tr.traverse(selfNode)
+      assertEquals(s"AST has the wrong number of $trgName", count, actualCount)
+      this
+    }
+
+    def traverse(pf: Pat): this.type = {
+      val tr = newTraverser(patToTotal(pf))
+      tr.traverse(selfNode)
+      this
+    }
+  }
+
+  class ClassDefAssertions(classDef: ClassDef)
+      extends AbstractIRNodeAssertions(classDef) {
+
+    protected def newTraverser(f: IRNode => Boolean): TestTraverser[ClassDef] = {
+      new TestTraverser[ClassDef](f) {
+        def baseTraverse(node: ClassDef): Unit = traverseClassDef(node)
+      }
+    }
+  }
+
+  class LinkedClassAssertions(linkedClass: LinkedClass)
+      extends AbstractIRNodeAssertions(linkedClass) {
+
+    protected def newTraverser(f: IRNode => Boolean): TestTraverser[LinkedClass] = {
+      new TestTraverser[LinkedClass](f) {
+        def baseTraverse(node: LinkedClass): Unit = {
+          for (memberDef <- node.fields)
+            traverseMemberDef(memberDef)
+          for (memberDef <- node.methods)
+            traverseMemberDef(memberDef.value)
+          for (memberDef <- node.exportedMembers)
+            traverseMemberDef(memberDef.value)
+          for (topLevelExportDef <- node.topLevelExports)
+            traverseTopLevelExportDef(topLevelExportDef.value)
+        }
+      }
+    }
+  }
+
+  abstract class TestTraverser[T](f: IRNode => Boolean) extends Traverser {
```

Comment:
Could we abstract this into a traverser and a finder? (traverser takes a `IRNode => Unit`, finder uses a traverser + control throwable). I think this will completely avoid the `finding` state.